### PR TITLE
1 - Transition Rules: Create API

### LINKED ISSUE #1347: Create Rule API
<This is intended as a guide when thinking about or creating issues for go-centrifuge>

#### Testing
<Describe required tests: Unit, Integration, Testworld or Functional>

#### Acceptance Criteria
<What is the acceptance criteria of this ticket?>

#### Dependencies and affected external components
<Outline any dependencies and components that could be affected by this>

### PR DESCRIPTION
**Pull request by vedhavyas**
Closes #1347

This PR adds the capability to add transition rules specific attributes in the document. Testworld tests help in using them.

### CODE REVIEW

File: documents/write_acls.go [hunk 1]
```diff
@@ -245,4 +247,98 @@ func (cd *CoreDocument) addNewTransitionRule(roleKey []byte, matchType coredocum
...
+		cd.addNewTransitionRule(
+			roleKey,
+			coredocumentpb.FieldMatchType_FIELD_MATCH_TYPE_EXACT,
+			f,
+			coredocumentpb.TransitionAction_TRANSITION_ACTION_EDIT)
+	}
+}
+
+// AddTransitionRuleForAttribute adds a new rule with key as fields for the role
+// Role must be present to create a rule.
```
Comment (mikiquantum) [inline documents/write_acls.go]:
```suggestion
// Role must be present to create a rule.
// Defaults to TransitionAction_TRANSITION_ACTION_EDIT
```

File: documents/write_acls.go [hunk 2]
```diff
@@ -245,4 +247,98 @@ func (cd *CoreDocument) addNewTransitionRule(roleKey []byte, matchType coredocum
 	}
 	cd.Document.TransitionRules = append(cd.Document.TransitionRules, rule)
 	cd.Modified = true
+	return rule
+}
+
+var attributeCompactPrefix = [...]byte{1, 0, 0, 0, 0, 0, 0, 28}
```
[the remark this answers is not in the document]
Comment (vedhavyas) [reply documents/write_acls.go]: MOre explained here - https://centrifuge.hackmd.io/TPZzPXTIRcarJEIOa_WZ7w?view#Attribute-compact-field-construction

File: pending/service.go [hunk 3]
```diff
@@ -278,3 +280,48 @@ func (s service) UpdateRole(ctx context.Context, docID, roleID []byte, collabs [
...
+	AttributeRules []AttributeRule `json:"attribute_rules"`
+}
+
+func (s service) AddTransitionRules(ctx context.Context, docID []byte, addRules AddTransitionRules) ([]*coredocumentpb.TransitionRule, error) {
+	accID, err := contextutil.AccountDID(ctx)
+	if err != nil {
+		return nil, contextutil.ErrDIDMissingFromContext
+	}
+
+	doc, err := s.pendingRepo.Get(accID[:], docID)
```
[the remark this answers is not in the document]
Comment (vedhavyas) [reply pending/service.go]: Document wouldn't be in pending state if it is already committed, right ?

File: documents/write_acls.go [hunk 4]
```diff
@@ -245,4 +247,98 @@ func (cd *CoreDocument) addNewTransitionRule(roleKey []byte, matchType coredocum
...
+
+// getAttributeField creates a compact property of the attribute key
+func getAttributeField(key AttrKey) []byte {
+	return append(attributeCompactPrefix[:], key[:]...)
+}
+
+// defaultRuleFieldProps are the fields that every collaborator should have rule set for to update a document.
+func defaultRuleFieldProps() map[string][]byte {
+	fields := [][]byte{
+		{1, 0, 0, 0, 0, 0, 0, 3},  // current_version
```
[the remark this answers is not in the document]
Comment (vedhavyas) [reply documents/write_acls.go]: Yes, these are used to create default rules for every role that has write access to attributes or any other fields that come next.
It is being used here - https://github.com/centrifuge/go-centrifuge/pull/1352/files#diff-b4a63c95b153e14cfaff40a71cdd8d8bR306

File: documents/write_acls.go [hunk 5]
```diff
@@ -245,4 +247,98 @@ func (cd *CoreDocument) addNewTransitionRule(roleKey []byte, matchType coredocum
...
+		return false
+	}
+
+	if _, ok := fieldMap[field]; !ok {
+		// not a match
+		return false
+	}
+
+	// delete the field from the map since the role is already present or we are going to add one to rule
+	delete(fieldMap, field)
```
[the remark this answers is not in the document]
Comment (vedhavyas) [reply documents/write_acls.go]: Updated the name of the function to represent the same

File: pending/service.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vedhavyas) [reply pending/service.go]: That would be deviating from the current flow.
A document either update or new one will be created using the following api that creates a new draft in to pending state.
https://app.swaggerhub.com/apis/centrifuge.io/cent-node/2.0.0#/Documents/create_document_v2

[the remark this answers is not in the document]
Comment (mikiquantum) [reply pending/service.go]: Agreed.

File: documents/write_acls.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (mikiquantum) [reply documents/write_acls.go]: Thanks, I am uncertain of what the function does return and in which cases, maybe add a block in the function comment to show in which cases the return is true or false, would be enough

**Comment by vedhavyas** [comment]
@mikiquantum will do!

**Pull request merged**